### ISSUE TITLE
`makeClassyPrisms` double definitions

### ISSUE THREAD

**Issue by ekmett** [ekmett/lens#865]
Something is rotten in Denmark.

```
newtype SDLException = SDLException String deriving Show
makeClassyPrisms ''SDLException
```

```
src/Engine/SDL.hs:20:1: error:
    Multiple declarations of ‘_SDLException’
    Declared at: src/Engine/SDL.hs:20:1
                 src/Engine/SDL.hs:20:1
   |
```

I don't use classy prisms very often, but this seems broken, as this is almost a textbook usecase.

**Comment by RyanGlScott** [comment]
What do you expect `makeClassyPrisms` to do here? The Haddocks seem to provide a loose specification by way of this example:

https://github.com/ekmett/lens/blob/60b89993903af95f138411a0721e952e84047a6c/src/Control/Lens/Internal/PrismTH.hs#L81-L102

Namely, it will generate prisms for both the name of the data type and all of its constructors. Since `SDLException` happens to be used as the name of both a data type and constructor, I suppose it makes sense that `makeClassyPrisms` would generate methods with duplicate names. Of course, it should probably throw a more informative error message instead, but that's a separate matter.

If you do expect this to work, what ought to be the specification for how `makeClassyPrisms` should handle duplicate names?

**Comment by ekmett** [comment]
I would expect that in the case where the only constructor name matches that we'd sensibly discard the one that pulls out the data type and provide just the one that provides the content.

This is exactly what we do in the 3 dozen hand-written versions of this that are in `Control.Exception.Lens`.

If that worked, then you could get to a pretty decent workflow for custom exceptions

```
data Shutdown = Shutdown deriving (Show,Exception)
makeClassyPrisms ''Shutdown
isntance AsShutdown SomeException where _Shutdown = exception
```

It would also be a path to generating rather than hand-rolling those 3-dozen instances.

**Comment by RyanGlScott** [comment]
Hm, OK. I suppose in this (somewhat contrived) example:

```hs
data A = A Int | B Bool
makeClassPrisms ''A
```

You would be OK with generating the following code?

```hs
class AsA s a | s -> a where
  _A :: Prism' s Int
  _B :: Prism' s Bool
```

It's also worth noting that since this class doesn't have a method of type `Prism' s A`, the usual approach for providing default implementations for `_A` and `_B` wouldn't work. This is easy enough to work around when generating the underlying `instance AsA A`, of course, but does mean that anyone who wishes to define their own `AsA` instance will have to implement each prism by hand.

**Comment by Taneb** [comment]
Poorly thought out suggestion: when there's an overlap, disambiguate with something like `_A'` or `_AType`

**Comment by ekmett** [comment]
Clearly the answer is `__A` =)

**Comment by RyanGlScott** [comment]
Your suggestion in https://github.com/ekmett/lens/issues/865#issuecomment-500967875 might be quite viable, actually. Since data constructor names aren't allowed to begin with underscores, there's no chance for a prism named `__A` to clash with the classy prism of some other data constructor named `_A`.

**Comment by RyanGlScott** [comment]
Of course, there would still be a question of whether the double-underscore-prefixed name should correspond to the datatype prism or the constructor prism. I'm guessing that you'd prefer it be the datatype name, judging from the fact that the code you wish to write in https://github.com/ekmett/lens/issues/865#issuecomment-500620210. It would also mean that the naming conventions we've already adopted in `Control.Exception.Lens` would continue to be consistent with what the TH machinery generates.

**Comment by ekmett** [comment]
That is indeed what I'd prefer.

**Comment by RyanGlScott** [comment]
I've implemented this approach in #869.

**Comment by RyanGlScott** [comment]
If we do adopt the approach in #869, one other thing to consider would be adding more prisms to `Control.Exception.Lens` to reflect this new naming convention. As one example, we could change `AsNonTermination` from this:

```hs
class AsNonTermination t where
  _NonTermination :: Prism' t ()
```

to this:

```hs
class AsNonTermination t where
  __NonTermination :: Prism' t NonTermination

  _NonTermination :: Prism' t ()
  _NonTermination = __NonTermination._NonTermination
```

Do you think this would be worthwhile? It would have the downside that any existing `AsNonTermination` instances would now need to implement `__AsNonTermination` to avoid warnings, but it would otherwise be a backwards-compatible change.
